js/interpreters.js

```diff
@@ -0,0 +1,132 @@
+/* Various interpreter to use with worksheet */
+import {referenceGrammar} from './ohm/reference.js';
+import {referenceSemantics} from './ohm/reference.js';
+import {commandGrammar} from './ohm/command.js';
+import {commandSemantics} from './ohm/command.js';
+
+class BasicInterpreter extends Object {
+    constructor(){
+        super();
+
+        this.interpret = this.interpret.bind(this);
+        this.matchAndInterpretCommand = this.matchAndInterpretCommand.bind(this);
+        this.matchAndInterpretReference = this.matchAndInterpretReference.bind(this);
+    }
+
+    interpret(instruction){
+        // NOTE: [source, target, command] convention
+        let [source, target, command] = instruction;
+        command = this.matchAndInterpretCommand(command);
+        const [name, args] = command;
+        source = this.matchAndInterpretReference(source);
+        target = this.matchAndInterpretReference(target);
+        return function(){
+            const exec = commandRegistry[name];
+            return exec(source, target, args);
+        };
+    }
+
+    matchAndInterpretCommand(s){
+        const m = commandGrammar.match(s);
+        if(m.failed()){
+            alert(`I cannot parse ${s}`);
+            // throw new Error(`NotUnderstood ${s}`);
+            return;
+        }
+        const result = commandSemantics(m).interpret();
+        if(!result){
+            alert(`I cannot interpret ${s}`);
+            // throw new Error(`NotKnown ${s}`);
+            return;
+        }
+        return result;
+    }
+
+    matchAndInterpretReference(s){
+        const m = referenceGrammar.match(s);
+        if(m.failed()){
+            alert(`I cannot parse ${s}`);
+            // throw new Error(`NotUnderstood ${s}`);
+            return;
+        }
+        const result = referenceSemantics(m).interpret();
+        if(!result){
+            alert(`I cannot interpret ${s}`);
+            // throw new Error(`NotKnown ${s}`);
+            return;
+        }
+        return result[0];
+    }
+}
+
+const copy = (source, target) => {
+    replace(source, target);
+}
+
+const replace = (source, target, d) => {
```

Comment:
Should we add some functionality to DataFrame in order to better enable this behavior? I feel like we should be able to take references to two DataFrame instances and just use normal frame methods to accomplish this (even if we haven't implemented the "right things" yet)

Reply:
i agree, my question is whether the api that the interpreter/worksheet uses should be at the selector or the dataFrame level. One thing I want to happen is live sub-frame/sub-sheet selector highlighting as you step through the commands. Something to think about